File: app/src/main/java/dev/radley/omgstarwars/view/CategoriesActivity.java
package dev.radley.omgstarwars.view;

import android.net.Uri;
import android.os.Bundle;
import android.view.Menu;
import android.view.View;
import android.widget.ImageView;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentStatePagerAdapter;
import androidx.lifecycle.ViewModelProviders;
import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.DrawableTransitionOptions;
import com.bumptech.glide.request.RequestOptions;
import com.google.android.material.appbar.AppBarLayout;
import com.google.android.material.tabs.TabLayout;

import org.jetbrains.annotations.NotNull;

import java.util.Objects;

import dev.radley.omgstarwars.R;
import dev.radley.omgstarwars.bundle.SearchExtras;
import dev.radley.omgstarwars.utilities.Constants;
import dev.radley.omgstarwars.viewmodels.CategoriesViewModel;
import dev.radley.omgstarwars.utilities.FormatUtils;

public class CategoriesActivity extends AppCompatActivity {

    private AppBarLayout appBarLayout;
    private AppBarLayout.OnOffsetChangedListener appBarOffsetListener;
    private SearchView searchView;
    private SearchView.SearchAutoComplete searchAutoComplete;
    private TabLayout tabLayout;
    private TabLayout.ViewPagerOnTabSelectedListener tabListener;
    private ViewPager viewPager;
    private ViewPager.OnPageChangeListener pageChangeListener;

    private static CategoriesViewModel viewModel;

    /**
     * Instatiate layouts and viewModel
     *
     * @param savedInstanceState Bundle
     */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_categories);
        setupToolbar();

        viewModel = ViewModelProviders.of(this).get(CategoriesViewModel.class);

        setupLayout();
    }

    /**
     * Start listeners
     */
    @Override
    public void onStart() {
        super.onStart();

        // add listeners
        viewPager.addOnPageChangeListener(pageChangeListener);
        tabLayout.addOnTabSelectedListener(tabListener);
        appBarLayout.addOnOffsetChangedListener(appBarOffsetListener);
    }

    /**
     * Stop listeners
     */
    @Override
    public void onStop() {
        super.onStop();

        // remove listeners
        viewPager.removeOnPageChangeListener(pageChangeListener);
        tabLayout.removeOnTabSelectedListener(tabListener);
        appBarLayout.removeOnOffsetChangedListener(appBarOffsetListener);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {

        getMenuInflater().inflate(R.menu.menu_categories, menu);
        searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        setupSearchView();

        return true;
    }

    /**
     * Setup searchview in toolbar
     */
    private void setupSearchView() {

        // add background color to searchView, only while expanded
        // (set this before calling first setIconified() so initial focus is established)
        //
        searchView.setOnQueryTextFocusChangeListener(new View.OnFocusChangeListener() {
            @Override
            public void onFocusChange(View v, boolean hasFocus) {

                if (hasFocus) {
                    searchView.setBackgroundColor(getColor(R.color.transparentPrimary));
                } else {
                    searchView.setBackground(null);
                }
            }
        });

        searchView.setQueryHint(getString(R.string.search_query_hint, viewModel.getCategory()));
        searchView.setIconified(false);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {

            @Override
            public boolean onQueryTextSubmit(String query) {

                // open search activity on submit
                startSearchActivity(query);
                return false;
            }

            @Override
            public boolean onQueryTextChange(String query) {

                // TODO could do autocomplete results
                // but that's a bit much for this exercise

                return true;
            }
        });
    }

    /**
     * setup toolbar
     */
    private void setupToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        Objects.requireNonNull(getSupportActionBar()).setDisplayShowTitleEnabled(false);
    }

    /**
     * Tell viewModel category has changed
     * update searchView hint and dispose any query text
     *
     * @param position
     */
    private void updateCategory(int position) {

        viewModel.setCategory(position);
        updateHeroImage();

        searchView.setQueryHint(getString(R.string.search_query_hint, viewModel.getCategory()));
        searchView.setQuery("", false);
    }

    /**
     * Instantiate views
     */
    private void setupLayout() {

        // setup view pager
        viewPager = findViewById(R.id.pager);
        PagerAdapter mPagerAdapter = new MyPagerAdapter(getSupportFragmentManager());
        viewPager.setAdapter(mPagerAdapter);
        pageChangeListener = new ViewPager.OnPageChangeListener() {

            @Override
            public void onPageSelected(int position) {
                updateCategory(position);
            }

            @Override
            public void onPageScrolled(int arg0, float arg1, int arg2) {
            }

            @Override
            public void onPageScrollStateChanged(int arg0) {
            }
        };

        // setup tabs
        tabLayout = findViewById(R.id.tabs);
        tabLayout.setupWithViewPager(viewPager);
        tabListener = new TabLayout.ViewPagerOnTabSelectedListener(viewPager) {

            // scroll to top on tab double-tap
            @Override
            public void onTabReselected(TabLayout.Tab tab) {

                CategoryFragment fragment = (CategoryFragment) Objects.requireNonNull(viewPager.getAdapter())
                        .instantiateItem(viewPager, viewPager.getCurrentItem());

                fragment.getRecyclerView().smoothScrollToPosition(0);
            }
        };

        // setup parallax hero header
        appBarLayout = findViewById(R.id.app_bar);
        appBarOffsetListener = (appBarLayout, verticalOffset) -> {

            // show background onionskin behind tabs when app bar is extended
            if (Math.abs(verticalOffset) < 24) {
                tabLayout.setBackgroundColor(getApplicationContext().getColor(R.color.transparentPrimaryDark));

            } else {
                tabLayout.setBackground(null);
            }
        };

        updateHeroImage();
    }

    /**
     * Clear searchView and start SearchActivity with query & category
     *
     * @param query String
     */
    private void startSearchActivity(String query) {

        searchView.setQuery("", false);
        searchView.setIconified(true);

        startActivity(SearchExtras.getIntent(this, FormatUtils.getTrimmedQuery(query), viewModel.getCategory()));
    }

    /**
     * Adapter for page tabs
     */
    public static class MyPagerAdapter extends FragmentStatePagerAdapter {

        private MyPagerAdapter(FragmentManager fragmentManager) {
            super(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @Override
        public int getCount() {
            return viewModel.getCategoryCount();
        }

        @NotNull
        @Override
        public Fragment getItem(int position) {
            return CategoryFragment.newInstance(viewModel.getCategory(position));
        }

        @Override
        public CharSequence getPageTitle(int position) {
            return viewModel.getCategoryTitle(position);
        }
    }

    /**
     * Update top hero background image when tab page changes
     */
    private void updateHeroImage() {

        // default image
        RequestOptions requestOptions = new RequestOptions()
                .placeholder(R.drawable.placeholder_hero);

        ImageView imageView = findViewById(R.id.hero_image);

        // load with fade in
        Glide.with(this)
                .setDefaultRequestOptions(requestOptions)
                .load(Uri.parse(Constants.HERO_ASSETS_PATH + viewModel.getCategory() + ".jpg"))
                .transition(DrawableTransitionOptions.withCrossFade())
                .into(imageView);
    }
}

File: OMGStarWars/app/src/main/java/dev/radley/omgstarwars/view/CategoriesActivity.kt
package dev.radley.omgstarwars.view

import android.graphics.Typeface
import android.os.Bundle
import android.view.Menu
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.SearchView
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentStatePagerAdapter
import androidx.viewpager.widget.ViewPager
import com.google.android.material.tabs.TabLayout
import dev.radley.omgstarwars.R
import dev.radley.omgstarwars.bundle.SearchExtras
import dev.radley.omgstarwars.models.Category
import dev.radley.omgstarwars.utilities.FormatUtils
import kotlinx.android.synthetic.main.activity_categories.*
import timber.log.Timber


class CategoriesActivity : AppCompatActivity() {

    private var categories = Category.categories
    private var category = categories[0]

    private lateinit var pagerAdapter: CategoriesPagerAdapter
    private lateinit var searchView: SearchView

    private val tabTypeface = Typeface.create("sans-serif-light",Typeface.NORMAL)
    private val tabTypefaceSelected = Typeface.create("sans-serif-black",Typeface.NORMAL)


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_categories)
        setupToolbar()
        setupLayout()
//        updateHeroImage()

        Timber.d("onCreate()")
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {

        menuInflater.inflate(R.menu.menu_categories, menu)
        searchView = menu!!.findItem(R.id.action_search).actionView as SearchView
        setupSearchView()
        return true
    }

    private fun setupSearchView() {

//        searchView.setOnQueryTextFocusChangeListener { _, hasFocus ->
//
//            if (hasFocus) {
//                searchView.setBackgroundColor(getColor(R.color.transparentPrimary))
//            } else {
//                searchView.background = null
//            }
//        }

        searchView.queryHint = getString(R.string.search_query_hint, category)
        searchView.isIconified = false

        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {

            override fun onQueryTextChange(newText: String): Boolean {
                return false
            }

            override fun onQueryTextSubmit(query: String): Boolean {
                startSearchActivity(query)
                return false
            }
        })
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar!!.setDisplayShowTitleEnabled(false)
    }

    var currentPage = 0

    private fun setupLayout() {
        pagerAdapter = CategoriesPagerAdapter(supportFragmentManager)

        viewPager.adapter = pagerAdapter
        viewPager.offscreenPageLimit = 6 // TODO add this line for recording video demo
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {

            override fun onPageSelected(position: Int) {
                updateCategory(position)

                val fragment = viewPager.adapter!!.instantiateItem(viewPager, viewPager.currentItem) as CategoryFragment

                if(position > currentPage) {
                    val controller = AnimationUtils.loadLayoutAnimation(fragment.activity, R.anim.category_layout_right_to_left_animation)
                    fragment.recyclerView.layoutAnimation = controller
                } else {
                    val controller = AnimationUtils.loadLayoutAnimation(fragment.activity, R.anim.category_layout_left_to_right_animation)
                    fragment.recyclerView.layoutAnimation = controller
                }

                currentPage = position
                fragment.recyclerView.startLayoutAnimation()
            }

            override fun onPageScrollStateChanged(state: Int) {}
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}
        })

        tabLayout.setupWithViewPager(viewPager)

        /* Tabs need to use a custom textview to be able to have:
         *
         * - margins for first & last tabs that match overall layout: 16 (sides) + 4 (card) = 20
         * - bold selected tab text
         *
         */
        for (i in 0 until tabLayout.tabCount) {

            val tab = tabLayout.getTabAt(i)
            if (tab != null) {

                val tabTextView = TextView(this)
                tab.customView = tabTextView

                tabTextView.typeface = tabTypeface
                tabTextView.layoutParams.width = ViewGroup.LayoutParams.WRAP_CONTENT
                tabTextView.layoutParams.height = ViewGroup.LayoutParams.WRAP_CONTENT

                tabTextView.text = tab.text
                tabTextView.isAllCaps = true
                tabTextView.textSize = 24f
                tabTextView.setTextColor(getColor(R.color.text_low_emphasis_color))

                val p = tabTextView.layoutParams as ViewGroup.MarginLayoutParams

                when (i) {
                    0 -> {
                        tabTextView.typeface = tabTypefaceSelected
                        tabTextView.setTextColor(getColor(R.color.tab_text_color_selected))
                        p.setMargins(28, 0, 8, 0)
                        tabTextView.requestLayout()
                    }
                    tabLayout.tabCount - 1 -> {
                        tabTextView.typeface = tabTypeface
                        tabTextView.setTextColor(getColor(R.color.text_low_emphasis_color))
                        p.setMargins(8, 0, 28, 0)
                        tabTextView.requestLayout()
                    }
                    else -> {
                        tabTextView.typeface = tabTypeface
                        tabTextView.setTextColor(getColor(R.color.text_low_emphasis_color))
                        p.setMargins(8, 0, 8, 0)
                        tabTextView.requestLayout()
                    }
                }
            }
        }

        tabLayout.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {

            override fun onTabSelected(tab: TabLayout.Tab) {

                // customize selected tab text style
                val vg = tabLayout.getChildAt(0) as ViewGroup
                val vgTab = vg.getChildAt(tab.position) as ViewGroup
                val tabChildsCount = vgTab.childCount

                for (i in 0 until tabChildsCount) {
                    val tabViewChild = vgTab.getChildAt(i)
                    if (tabViewChild is TextView) {
                        tabViewChild.typeface = tabTypefaceSelected
                        tabViewChild.setTextColor(getColor(R.color.tab_text_color_selected))
                    }
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab?) {

                // customize unselected tab text style
                tab?.let {

                    val vg = tabLayout.getChildAt(0) as ViewGroup
                    val vgTab = vg.getChildAt(it.position) as ViewGroup
                    val tabChildsCount = vgTab.childCount

                    for (i in 0 until tabChildsCount) {
                        val tabViewChild = vgTab.getChildAt(i)
                        if (tabViewChild is TextView) {
                            tabViewChild.typeface = tabTypeface
                            tabViewChild.setTextColor(getColor(R.color.text_low_emphasis_color))
                        }
                    }
                }

                val fragment = viewPager.adapter!!.instantiateItem(viewPager, viewPager.currentItem) as CategoryFragment
                fragment.recyclerView.smoothScrollToPosition(0)
            }

            override fun onTabReselected(tab: TabLayout.Tab?) {

                val fragment = viewPager.adapter!!.instantiateItem(viewPager, viewPager.currentItem) as CategoryFragment
                fragment.recyclerView.smoothScrollToPosition(0)
            }
        })
    }

    inner class CategoriesPagerAdapter(fragmentManager: FragmentManager) :
            FragmentStatePagerAdapter(fragmentManager, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT) {

        override fun getCount() = categories.size
        override fun getPageTitle(position: Int) = categories[position].capitalize()
        override fun getItem(position: Int): Fragment = CategoryFragment.newInstance(categories[position])
    }

    private fun startSearchActivity(query: String) {
        searchView.setQuery("", false)
        searchView.isIconified = true

        startActivity(SearchExtras.getIntent(this, FormatUtils.getTrimmedQuery(query), category))
    }

    private fun updateCategory(position: Int) {
        category = categories[position]
        searchView.queryHint = getString(R.string.search_query_hint, category)
    }
}